### ISSUE TITLE
Stream#map not stacksafe

### ISSUE THREAD

**Issue by TimWSpence** [typelevel/fs2#2353]
```
  test("boom") {
    def loop(str: Stream[Pure, Int], i: Int): Stream[Pure, Int] =
      if (i == 0) str else loop(str.map((x: Int) => x + 1), i - 1)

    loop(Stream.emit(1), 10000).compile.toList //
  }
```

stack-overflows for me. It looks at first glance like using `cats.data.AndThen` [here](https://github.com/typelevel/fs2/blob/dee95aa1f4db8d31f01e394e0595a4dcf7922604/core/shared/src/main/scala/fs2/Pull.scala#L1269) might fix it but I haven't tried it to confirm.

Happy to give it a go though!

**Comment by TimWSpence** [comment]
Using `AndThen` did seem to fix it so I created https://github.com/typelevel/fs2/pull/2354

**Comment by kubukoz** [comment]
Does this affect effectful streams too?

If it doesn't, I'd be wary of adding `AndThen` for both, especially if it incurs a performance hit on the already stacksafe effectful ones. Disregard this comment if it's the same with e.g. `IO` :P

**Comment by djspiewak** [comment]
This definitely applies to `IO` as well. The stack unsafety is in `Function1`

**Comment by diesalbla** [comment]
> This definitely applies to IO as well. The stack unsafety is in `Function1`

So, should we then make the code of `Stream` and `IO` to safeguard against that? Or at least, to avoid composition in our code (the problem here stems from the interpretation of `map`).

I wonder if the stack-safe solution (I am guessing passing the left-hand-sides of composition in continuation) could be implemented on the Scala library itself

**Comment by TimWSpence** [comment]
>  wonder if the stack-safe solution (I am guessing passing the left-hand-sides of composition in continuation) could be implemented on the Scala library itself

Yeah it _really_ feels like we should move it into the standard library implementation

**Comment by mpilquist** [comment]
Fixed by #2354
